src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -671,28 +672,51 @@ private async Task EnsureNuGetAndVsProjectAdapterCacheAsync()
             {
                 await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
 
-                if (!_cacheInitialized && await IsSolutionOpenAsync())
+                IVsSolution ivsSolution = await _asyncVSSolution.GetValueAsync();
+                if (!_cacheInitialized && IsSolutionOpenFromVSSolution(ivsSolution))
                 {
                     try
                     {
-                        var dte = await _dte.GetValueAsync();
-
-                        foreach (var project in await EnvDTESolutionUtility.GetAllEnvDTEProjectsAsync(dte))
+                        if (await _featureFlagService.IsFeatureEnabledAsync(NuGetFeatureFlagConstants.NuGetSolutionCacheInitilization))
                         {
-                            try
+                            foreach (var hierarchy in VsHierarchyUtility.GetAllLoadedProjects(ivsSolution))
                             {
-                                var vsProjectAdapter = await _vsProjectAdapterProvider.CreateAdapterForFullyLoadedProjectAsync(project);
-                                await AddVsProjectAdapterToCacheAsync(vsProjectAdapter);
+                                try
+                                {
+                                    var vsProjectAdapter = await _vsProjectAdapterProvider.CreateAdapterForFullyLoadedProjectAsync(hierarchy);
+                                    await AddVsProjectAdapterToCacheAsync(vsProjectAdapter);
+                                }
+                                catch (Exception e)
+                                {
+                                    // Ignore failed projects.
+                                    _logger.LogWarning($"The project {VsHierarchyUtility.GetProjectPath(hierarchy)} failed to initialize as a NuGet project.");
+                                    _logger.LogError(e.ToString());
+                                }
+
+                                // Consider that the cache is initialized only when there are any projects to add.
+                                _cacheInitialized = true;
                             }
-                            catch (Exception e)
+                        }
+                        {
+                            var dte = await _dte.GetValueAsync();
+
+                            foreach (var project in await EnvDTESolutionUtility.GetAllEnvDTEProjectsAsync(dte))
```

Comment:
Why the double-walk of the solution? I expected this to be removed.

Reply:
I meant to have an else here. 
I want to have a feature flag to fallback to the old behavior in case I missed something. 

The consequences of a potential bug here are the project(s) not initializing correctly and NuGet simply not working correctly.

I think the change is safe, but it'd be easier to argue for it in case it doesn't make it into an insertion by Wednesday.